controllers/vaultsecret_controller.go

```diff
@@ -0,0 +1,224 @@
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	ctx = context.WithValue(ctx, "kubeClient", r.GetClient())
+
+	err = r.manageReconcileLogic(ctx, instance)
+	if err != nil {
+		r.Log.Error(err, "unable to complete reconcile logic", "instance", instance)
+		return r.ManageError(ctx, instance, err)
+	}
+
+	//we reschedule the next reconcile at the time in the future corresponding to
+	nextSchedule := time.Until(instance.Status.LastVaultSecretUpdate.Add(instance.Spec.RefreshPeriod.Duration))
+	if nextSchedule > 0 {
+		return r.ManageSuccessWithRequeue(ctx, instance, nextSchedule)
+	} else {
+		return r.ManageSuccessWithRequeue(ctx, instance, time.Second)
+	}
+
+}
+
+func (r *VaultSecretReconciler) formatK8sSecret(instance *redhatcopv1alpha1.VaultSecret, data interface{}) (*corev1.Secret, error) {
+
+	stringData := make(map[string]string)
+	for k, v := range instance.Spec.TemplatizedK8sSecret.StringData {
+
+		tpl, err := template.New("").Funcs(utilstemplates.AdvancedTemplateFuncMap(r.GetRestConfig(), r.Log)).Parse(v)
+		if err != nil {
+			r.Log.Error(err, "unable to create template", "instance", instance)
+			return nil, err
+		}
+
+		var b bytes.Buffer
+		err = tpl.Execute(&b, data)
+		if err != nil {
+			r.Log.Error(err, "unable to execute template", "instance", instance)
+			return nil, err
+		}
+
+		stringData[k] = b.String()
+	}
+
+	k8sSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        instance.Spec.TemplatizedK8sSecret.Name,
+			Namespace:   instance.Namespace,
+			Annotations: instance.Spec.TemplatizedK8sSecret.Annotations,
+			Labels:      instance.Spec.TemplatizedK8sSecret.Labels,
+		},
+		StringData: stringData,
+		Type:       corev1.SecretType(instance.Spec.TemplatizedK8sSecret.Type),
+	}
+
+	ctrl.SetControllerReference(instance, k8sSecret, r.GetScheme())
+
+	return k8sSecret, nil
+}
+
+func (r *VaultSecretReconciler) manageReconcileLogic(ctx context.Context, instance *redhatcopv1alpha1.VaultSecret) error {
+
+	mergedMap := make(map[string]interface{})
+	for _, kvSecret := range instance.Spec.KVSecrets {
+		vaultClient, err := kvSecret.Authentication.GetVaultClient(ctx, instance.Namespace)
+		if err != nil {
+			r.Log.Error(err, "unable to create vault client", "instance", instance)
+			return err
+		}
+
+		ctx = context.WithValue(ctx, "vaultClient", vaultClient)
+		vaultEndpoint := vaultutils.NewVaultEndpointObj(&kvSecret)
+
+		data, ok, _ := vaultEndpoint.Read(ctx)
+		if !ok {
+			return errors.New("unable to read Key/Value Secret from Vault for " + kvSecret.GetPath())
+		}
+
+		mergedMap[kvSecret.Name] = data
+	}
+
+	k8sSecret, err := r.formatK8sSecret(instance, mergedMap)
+	if err != nil {
+		r.Log.Error(err, "unable to format k8s secret", "instance", instance)
+		return err
+	}
+
+	existingK8sSecret := &corev1.Secret{}
```

Comment:
please use the createOrUpdate function from utils. This will also take care of controller reference